app-backend/batch/src/main/scala/com/azavea/rf/batch/ast/package.scala

```diff
@@ -0,0 +1,104 @@
+package com.azavea.rf.batch
+
+import java.util.UUID
+
+import cats.data.NonEmptyList
+import cats.data.Validated._
+import cats.implicits._
+import com.azavea.rf.tool.ast._
+import com.azavea.rf.tool.ast.MapAlgebraAST._
+import com.azavea.rf.tool.eval._
+import com.azavea.rf.tool.eval.Interpreter.Interpreted
+import geotrellis.raster.{MultibandTile, Tile}
+import geotrellis.spark._
+import geotrellis.spark.io._
+import geotrellis.spark.io.s3._
+import geotrellis.spark.render._
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.RDD
+
+// --- //
+
+package object ast {
+
+  /** Perform a binary operation on RDDs, while preserving any metadata they had. */
+  private def binary(
+    f: (RDD[(SpatialKey, Tile)], RDD[(SpatialKey, Tile)]) => RDD[(SpatialKey, Tile)],
+    tlr0: TileLayerRDD[SpatialKey],
+    tlr1: TileLayerRDD[SpatialKey]
+  ): TileLayerRDD[SpatialKey] = {
+    TileLayerRDD(f(tlr0, tlr1), tlr0.metadata.combine(tlr1.metadata))
+  }
+
+  /** Evaluate an AST of RDD Sources. Assumes that the AST's
+    * [[NodeMetadata]] has already been replaced, if applicable.
+    */
+  def interpretRDD(
+    ast: MapAlgebraAST,
+    sourceMapping: Map[UUID, RFMLRaster],
+    zoom: Int,
+    ingestLocs: Map[UUID, String]
+  )(implicit sc: SparkContext): Interpreted[TileLayerRDD[SpatialKey]] = {
+
+    /* Perform Map Algebra over a validated RDD-filled AST */
+    def eval(
+      ast: MapAlgebraAST,
+      rdds: Map[UUID, TileLayerRDD[SpatialKey]]
+    ): TileLayerRDD[SpatialKey] = ast match {
+      case Source(id, _) => rdds(id)
+      case Addition(args, _, _) =>
+        args.map(eval(_, rdds)).reduce((acc,r) => binary({_ + _}, acc, r))
+      case Subtraction(args, _, _) =>
+        args.map(eval(_, rdds)).reduce((acc,r) => binary({_ - _}, acc, r))
+      case Multiplication(args, _, _) =>
+        args.map(eval(_, rdds)).reduce((acc,r) => binary({_ * _}, acc, r))
+      case Division(args, _, _) =>
+        args.map(eval(_, rdds)).reduce((acc,r) => binary({_ / _}, acc, r))
+      case Classification(arg :: _, _, _, classMap) =>
+        eval(arg, rdds).withContext(_.color(classMap.toColorMap))
+      case _ => ???
```

Comment:
At least to throw an exception (that would be at least human readable if smth would happen)